internal/oauthutil/token.go

```
@@ -0,0 +1,237 @@
+	if err != nil {
+		return err
+	}
+	i, err := n.Int64()
+	if err != nil {
+		return err
+	}
+	if i > math.MaxInt32 {
+		i = math.MaxInt32
+	}
+	*e = expirationTime(i)
+	return nil
+}
+
+type AuthStyle int
+
+const (
+	AuthStyleInParams AuthStyle = 1
+	AuthStyleInHeader AuthStyle = 2
+)
+
+// newTokenRequest returns a new *http.Request to retrieve a new token
+// from tokenURL using the provided clientID, clientSecret, and POST
+// body parameters.
+//
+// If AuthStyleInParams is true, the provided values will be encoded in the POST body.
+func newTokenRequest(oauthCtx OauthContext, refreshToken string, authStyle AuthStyle) (*http.Request, error) {
+	v := url.Values{}
+	if authStyle == AuthStyleInParams {
+		v.Set("client_id", oauthCtx.ClientID)
+		v.Set("client_secret", oauthCtx.ClientSecret)
+		v.Set("grant_type", "refresh_token")
+		v.Set("refresh_token", refreshToken)
+	}
+
+	req, err := http.NewRequest("POST", oauthCtx.Endpoint.TokenURL, strings.NewReader(v.Encode()))
+
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	if authStyle == AuthStyleInHeader {
+		req.SetBasicAuth(url.QueryEscape(oauthCtx.ClientID), url.QueryEscape(oauthCtx.ClientSecret))
+	}
+	return req, nil
+}
+
+func RetrieveToken(doer httpcli.Doer, oauthCtx OauthContext, refreshToken string, authStyle AuthStyle) (*Token, error) {
+	req, err := newTokenRequest(oauthCtx, refreshToken, authStyle)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := doTokenRoundTrip(doer, req)
+	if err != nil {
+		return nil, errors.Wrap(err, "do token round trip")
+	}
+
+	if token != nil && token.RefreshToken == "" {
+		token.RefreshToken = refreshToken
+	}
+	return token, err
+}
+
+func doTokenRoundTrip(doer httpcli.Doer, req *http.Request) (*Token, error) {
+	r, err := doer.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "do request")
+	}
+
+	body, err := io.ReadAll(io.LimitReader(r.Body, 1<<20))
+	r.Body.Close()
```

Might want to check the error here too

---

@ryanslade , I am wondering if this format would be acceptable for SG codebase

````
	body, err := io.ReadAll(io.LimitReader(r.Body, 1<<20))
	defer func(Body io.ReadCloser) {
		err := Body.Close()
		if err != nil {
			logger.Error("oauth2: error closing body", log.Error(err))
		}
	}(r.Body)

---

I'm leaning towards just ignoring the `Close` error here.

---

I'm ok with just ignoring the error like Joe said, we have many places in the codebase where we do something like that:
https://sourcegraph.sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+_+%3D+%5Cw.Close%28%29&patternType=regexp